### PR TITLE
[#11469] Browser bookmarks

### LINKED ISSUE #11469: Browser bookmarks
# Feature Issue

image

### User Story

As a user i would like to have bookmarks in the browser

### PR DESCRIPTION
**Pull request by flexsurfer**
fixes #11469

this PR doesn't include:
- search recently opened and bookmarks
- tabs design , animation and screenshots
- persist tabs

issue https://github.com/status-im/status-react/issues/11538

### CODE REVIEW

File: src/status_im/ui/screens/browser/options/views.cljs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            :icon                :main-icons/add
+            :on-press            #(hide-sheet-and-dispatch [:navigate-to :empty-tab])}]
+          [quo/list-item
+           {:theme               :accent
+            :title               (if fav? (i18n/label :t/remove-favourite) (i18n/label :t/favourite))
+            :accessibility-label :add-remove-fav
+            :icon                (if fav? :main-icons/delete :main-icons/favourite)
+            :on-press            #(hide-sheet-and-dispatch (if fav?
+                                                             [:browser/delete-bookmark url]
+                                                             [:navigate-to :new-bookmark {:url url}]))}]
```
Comment (Ferossgp) [inline src/status_im/ui/screens/browser/options/views.cljs]: Looks like here the name is missing, which should autofill the default value in input on the new-bookmark screen?

**Review by Ferossgp — APPROVED** [review]
I found it strange that the tab state is killed right after the switch to a different, disabling the ability to do multitasking. And there are more small issues when navigating between tabs, but I think QA/Design review can better describe them.

File: src/status_im/ui/screens/browser/options/views.cljs (hunk 1 above)
Comment (flexsurfer) [reply src/status_im/ui/screens/browser/options/views.cljs]: if there is no name , it means its a new bookmark

Comment (Ferossgp) [reply src/status_im/ui/screens/browser/options/views.cljs]: Yes, I mean that in other browsers the default name uses the site title, while in our case it is empty

Comment (flexsurfer) [reply src/status_im/ui/screens/browser/options/views.cljs]: oh ok, good point

**Comment by qoqobolo** [comment]
### ISSUE 1: Error "v must satisfy IVector" if open some website or dapp and quickly tap on the back button several times

Scenario 1:
- Send link to any chat
- Tap and open it in Status
- Quickly tap back button several times until you are on the browser screen

Scenario 2:
- Open some dapp
- Several times follow links in this app
- Quickly tap back button several times until you are on the browser screen

Android, iOS

[geth.log](https://github.com/status-im/status-react/files/5672610/geth.log)
[Status.log](https://github.com/status-im/status-react/files/5672611/Status.log)

ezgif com-gif-maker (13)
ezgif com-gif-maker (14)

**Comment by flexsurfer** [comment]
thanks @qoqobolo  fixed

**Comment by qoqobolo** [comment]
### ISSUE 2: No Сancel or cross button for the URL input field

Currently there is no way to delete the entered URL and close a website with one button. Need to remove it manually or tap back button or close all tabs (Tabs view).

**_Case 1:_**
Screenshot 2020-12-10 at 15 53 44

_____________
**_Case 2_** (iOS has cross button here):
Screenshot 2020-12-10 at 15 54 03

**Comment by flexsurfer** [comment]
thanks ISSUE 2
case1 as designed
case2 search isn't implemented in this PR so no cancel button, it works in pair with the search

**Comment by errorists** [comment]
list of improvements
• rename `New Tab` to `Empty tab`, 
• the no.1 way to improve the UX is to bring favicons into the list
• let's not reorder the list when opening tabs, each tab should remain at its index.
• Dark Grey colour for the (X) icons
• for some reason when opening the tabs view from an empty tab I get the transition but not when I navigate from a website
• let's rename the `Favourite` button to `Add favourite`
• when you add a new Favourite, the name input field should already be filled with the website title attribute. so if you favourite dap.ps it would be `Discover Dapps | Status`
• why can't I open multiple instances of empty tabs? It's a bit weird coming from other browsers, to find empty tab being permanently opened and only a single instance of it. It would be better if we could have multiple instances of empty tabs, and if you have only one left on the list it couldn't be closed.

**Comment by flexsurfer** [comment]
- [x] cancel for input
- [x] rename New Tab to Empty tab,
- [x] show favicons.ico
- [x] let's not reorder the list when opening tabs, each tab should remain at its index.
- [x] Dark Grey colour for the (X) icons
- [x] let's rename the Favourite button to Add favourite
- [x] when you add a new Favourite, the name input field should already be filled with the website title attribute. so if you favourite dap.ps it would be Discover Dapps | Status


- [ ] the no.1 way to improve the UX is to bring favicons into the list / 
- [x] for some reason when opening the tabs view from an empty tab I get the transition but not when I navigate from a website 

•  browser screen doesn't have transition because of https://github.com/status-im/status-react/issues/10189 

- [ ] why can't I open multiple instances of empty tabs? It's a bit weird coming from other browsers, to find empty tab being permanently opened and only a single instance of it. It would be better if we could have multiple instances of empty tabs, and if you have only one left on the list it could be closed. 

•  agree its weird, but it will require to store empty tabs :( and manage them, so i decided to move it as a separate feature, if we really want it

**Comment by errorists** [comment]
@flexsurfer 
> case1 as designed

is it? then the designer is wrong :) there should be a clear input icon. The 'Cancel' button is likewise necessary to dismiss the input, it's not only for search. otherwise I can't dismiss the input once I open it.

Screenshot 2020-12-11 at 11 10 32

**Comment by flexsurfer** [comment]
thanks @errorists , @qoqobolo ready for review

**Comment by churik** [comment]
@flexsurfer can I ask you to add accessibility ids to following items if possible?
Appium 2020-12-11 14-12-25
Appium 2020-12-11 14-13-42
Appium 2020-12-11 14-16-50
Android Emulator - my_emu_nex:5554 2020-12-11 14-18-27 
Appium 2020-12-11 14-25-36

**Comment by qoqobolo** [comment]
### ISSUE 3: Crash when trying to edit favorite bookmark, if it was added to favorites with default name

**_Steps:_**

- Open browser
- Enter URL (i.e. github.com)
- Three dots at the bottom right -> Add favorite
- Tap Add favorite
- Tap Tabs icon
- Tap Empty tab
- Long press on added bookmark
- Edit

Android, iOS

[Status-debug-logs.zip](https://github.com/status-im/status-react/files/5679360/Status-debug-logs.4.zip)


ezgif com-gif-maker (15)

**Comment by flexsurfer** [comment]
thanks @qoqobolo fixed, @churik  added

**Comment by qoqobolo** [comment]
@flexsurfer great work, thanks!
Ready to merge.

**Pull request merged**